File: src/settings/reader.ts
import { readFileSync } from "node:fs";
import type { ResolvedEditor } from "../editors/types.js";

export interface EditorSettings {
  editor: ResolvedEditor;
  settings: Record<string, unknown>;
  raw: string;
}

export function readSettings(editor: ResolvedEditor): EditorSettings {
  if (!editor.settingsExist) {
    return { editor, settings: {}, raw: "{}" };
  }

  try {
    const raw = readFileSync(editor.settingsPath, "utf-8");
    const settings = JSON.parse(raw) as Record<string, unknown>;
    return { editor, settings, raw };
  } catch {
    return { editor, settings: {}, raw: "{}" };
  }
}

export function readAllSettings(
  editors: ResolvedEditor[]
): EditorSettings[] {
  return editors.map(readSettings);
}

File: src-tauri/src/settings/reader.rs
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use crate::editors::ResolvedEditor;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EditorSettings {
    pub editor: ResolvedEditor,
    pub settings: Map<String, Value>,
    pub raw: String,
}

pub fn read_settings(editor: &ResolvedEditor) -> EditorSettings {
    let path = std::path::Path::new(&editor.settings_path);
    if !path.exists() {
        return EditorSettings {
            editor: editor.clone(),
            settings: Map::new(),
            raw: "{}".to_string(),
        };
    }

    let raw = std::fs::read_to_string(path).unwrap_or_else(|_| "{}".to_string());
    let value: Value = serde_json::from_str(&raw).unwrap_or_else(|_| Value::Object(Map::new()));

    let settings = match value {
        Value::Object(map) => map,
        _ => Map::new(),
    };

    EditorSettings {
        editor: editor.clone(),
        settings,
        raw,
    }
}

pub fn read_all_settings(editors: &[ResolvedEditor]) -> Vec<EditorSettings> {
    editors.iter().map(read_settings).collect()
}
